test/es-module/test-esm-loader-hooks.mjs

```diff
@@ -1,83 +1,26 @@
-    assert.strictEqual(specifier, originalSpecifier);
-    // Ensure `context` has all and only the properties it's supposed to
-    assert.deepStrictEqual(Object.keys(context), [
-      'conditions',
-      'importAssertions',
-      'parentURL',
-    ]);
-    assert.ok(Array.isArray(context.conditions));
-    assert.deepStrictEqual(context.importAssertions, importAssertions);
-    assert.strictEqual(context.parentURL, parentURL);
-    assert.strictEqual(typeof defaultResolve, 'function');
-
-    return {
-      format: suggestedFormat,
-      shortCircuit: true,
-      url: resolvedURL,
-    };
-  }
-
-  function load(resolvedURL, context, defaultLoad) {
-    assert.strictEqual(resolvedURL, resolvedURL);
-    assert.ok(new URL(resolvedURL));
-    // Ensure `context` has all and only the properties it's supposed to
-    assert.deepStrictEqual(Object.keys(context), [
-      'format',
-      'importAssertions',
+import { spawnPromisified } from '../common/index.mjs';
+import * as fixtures from '../common/fixtures.mjs';
+import assert from 'node:assert';
+import { execPath } from 'node:process';
+import { describe, it } from 'node:test';
+
+describe('Loader hooks', () => {
+  it('are called with all expected arguments', async () => {
+    const { code, signal, stdout, stderr } = await spawnPromisified(execPath, [
+      '--no-warnings',
+      '--experimental-loader',
+      fixtures.fileURL('/es-module-loaders/hooks-input.mjs'),
+      fixtures.path('/es-modules/json-modules.mjs'),
     ]);
-    assert.strictEqual(context.format, suggestedFormat);
-    assert.deepStrictEqual(context.importAssertions, importAssertions);
-    assert.strictEqual(typeof defaultLoad, 'function');
-
-    // This doesn't matter (just to avoid errors)
-    return {
-      format: 'module',
-      shortCircuit: true,
-      source: '',
-    };
-  }
-
-  const customLoader = [
-    {
-      exports: {
-        // Ensure ESMLoader actually calls the custom hooks
-        resolve: mustCall(resolve),
-        load: mustCall(load),
-      },
-      url: import.meta.url,
-    },
-  ];
-
-  esmLoader.addCustomLoaders(customLoader);
 
-  // Manually trigger hooks (since ESMLoader is not actually running)
-  const job = await esmLoader.getModuleJob(
-    originalSpecifier,
-    parentURL,
-    importAssertions,
-  );
-  await job.modulePromise;
-}
+    assert.strictEqual(stderr, '');
+    assert.strictEqual(code, 0);
+    assert.strictEqual(signal, null);
+
+    const lines = stdout.split('\n');
+    assert.match(lines[0], /{"url":"file:\/\/\/.*\/json-modules\.mjs","format":"test","shortCircuit":true}/);
+    assert.match(lines[1], /{"source":{"type":"Buffer","data":\[.*\]},"format":"module","shortCircuit":true}/);
+    assert.match(lines[2], /{"url":"file:\/\/\/.*\/experimental\.json","format":"test","shortCircuit":true}/);
+    assert.match(lines[3], /{"source":{"type":"Buffer","data":\[.*\]},"format":"json","shortCircuit":true}/);
```

Comment:
Wouldn't it be clearer to `JSON.parse` each line and use `assert.deepStrictEqual`?

Reply:
The lines contain file paths specific to the system on which this is being run, and a long string of integers for the `Buffer` contents (see the `.*`s in the regexes). If I parsed the JSON I’d have to run separate `assert.strictEqual` and `assert.match` comparisons for every property, depending on whether it needs substitution or not. This is much shorter and faster.